File: electron/main/terminal-transcript-buffer.js
class TerminalTranscriptBuffer {
  constructor(seedText = "", maxLength = 160000) {
    this.maxLength = maxLength;
    this.committedText = "";
    this.currentLine = [];
    this.cursorColumn = 0;
    this.replaceSeed(seedText);
  }

  get renderedText() {
    return this.committedText + this.currentLine.join("");
  }

  consume(rawText) {
    this.apply(rawText);
    this.trimIfNeeded();
    return this.renderedText;
  }

  replaceSeed(text) {
    const newlineIndex = text.lastIndexOf("\n");

    if (newlineIndex >= 0) {
      this.committedText = text.slice(0, newlineIndex + 1);
      this.currentLine = Array.from(text.slice(newlineIndex + 1));
    } else {
      this.committedText = "";
      this.currentLine = Array.from(text);
    }

    this.cursorColumn = this.currentLine.length;
  }

  apply(rawText) {
    const chars = Array.from(rawText);
    let index = 0;

    while (index < chars.length) {
      const char = chars[index];

      if (char === "\u001b") {
        index = this.parseEscapeSequence(chars, index);
        continue;
      }

      switch (char) {
        case "\n":
          this.committedText += `${this.currentLine.join("")}\n`;
          this.currentLine = [];
          this.cursorColumn = 0;
          break;
        case "\r":
          this.cursorColumn = 0;
          break;
        case "\b":
        case "\u007f":
          if (this.cursorColumn > 0) {
            this.cursorColumn -= 1;

            if (this.cursorColumn < this.currentLine.length) {
              this.currentLine.splice(this.cursorColumn, 1);
            }
          }
          break;
        case "\t":
          this.write(char);
          break;
        default:
          if (!isControlCharacter(char)) {
            this.write(char);
          }
          break;
      }

      index += 1;
    }
  }

  parseEscapeSequence(chars, startIndex) {
    const nextIndex = startIndex + 1;
    if (nextIndex >= chars.length) {
      return startIndex + 1;
    }

    switch (chars[nextIndex]) {
      case "[":
        return this.parseCSISequence(chars, nextIndex + 1);
      case "]":
        return this.skipOSCSequence(chars, nextIndex + 1);
      default:
        return nextIndex + 1;
    }
  }

  parseCSISequence(chars, startIndex) {
    let index = startIndex;
    let parameterString = "";

    while (index < chars.length) {
      const char = chars[index];
      const code = char.charCodeAt(0);

      if (code >= 0x40 && code <= 0x7e) {
        this.applyCSI(char, parameterString);
        return index + 1;
      }

      parameterString += char;
      index += 1;
    }

    return index;
  }

  skipOSCSequence(chars, startIndex) {
    let index = startIndex;

    while (index < chars.length) {
      const char = chars[index];

      if (char === "\u0007") {
        return index + 1;
      }

      if (char === "\u001b" && index + 1 < chars.length && chars[index + 1] === "\\") {
        return index + 2;
      }

      index += 1;
    }

    return index;
  }

  applyCSI(finalChar, parameterString) {
    const parameters = parameterString.split(";").map((value) => Number.parseInt(value, 10) || 0);

    switch (finalChar) {
      case "m":
        return;
      case "K":
        this.clearLine(parameters[0] ?? 0);
        return;
      case "J":
        this.clearScreen(parameters[0] ?? 0);
        return;
      case "C":
        this.moveCursorForward(Math.max(parameters[0] ?? 1, 1));
        return;
      case "D":
        this.moveCursorBackward(Math.max(parameters[0] ?? 1, 1));
        return;
      case "G":
        this.moveCursorToColumn(Math.max((parameters[0] ?? 1) - 1, 0));
        return;
      case "P":
        this.deleteCharacters(Math.max(parameters[0] ?? 1, 1));
        return;
      case "X":
        this.eraseCharacters(Math.max(parameters[0] ?? 1, 1));
        return;
      default:
        return;
    }
  }

  write(char) {
    if (this.cursorColumn < this.currentLine.length) {
      this.currentLine[this.cursorColumn] = char;
    } else {
      while (this.currentLine.length < this.cursorColumn) {
        this.currentLine.push(" ");
      }

      this.currentLine.push(char);
    }

    this.cursorColumn += 1;
  }

  clearLine(mode) {
    switch (mode) {
      case 1: {
        const upperBound = Math.min(this.cursorColumn, this.currentLine.length);
        for (let index = 0; index < upperBound; index += 1) {
          this.currentLine[index] = " ";
        }
        return;
      }
      case 2:
        this.currentLine = [];
        this.cursorColumn = 0;
        return;
      default:
        if (this.cursorColumn < this.currentLine.length) {
          this.currentLine.splice(this.cursorColumn);
        }
    }
  }

  clearScreen(mode) {
    if (mode === 2 || mode === 3) {
      this.committedText = "";
      this.currentLine = [];
      this.cursorColumn = 0;
    }
  }

  moveCursorForward(amount) {
    this.cursorColumn += amount;
  }

  moveCursorBackward(amount) {
    this.cursorColumn = Math.max(this.cursorColumn - amount, 0);
  }

  moveCursorToColumn(column) {
    this.cursorColumn = Math.max(column, 0);
  }

  deleteCharacters(count) {
    if (this.cursorColumn < this.currentLine.length) {
      const upperBound = Math.min(this.cursorColumn + count, this.currentLine.length);
      this.currentLine.splice(this.cursorColumn, upperBound - this.cursorColumn);
    }
  }

  eraseCharacters(count) {
    if (this.cursorColumn >= this.currentLine.length) {
      return;
    }

    const upperBound = Math.min(this.cursorColumn + count, this.currentLine.length);
    for (let index = this.cursorColumn; index < upperBound; index += 1) {
      this.currentLine[index] = " ";
    }
  }

  trimIfNeeded() {
    if (this.renderedText.length > this.maxLength) {
      this.replaceSeed(this.renderedText.slice(-this.maxLength));
    }
  }

  static visibleText(rawText) {
    return new TerminalTranscriptBuffer("", Math.max(rawText.length * 2, 4096)).consume(rawText);
  }
}

function isControlCharacter(char) {
  const code = char.codePointAt(0) || 0;
  return (code >= 0 && code <= 0x1f) || code === 0x7f;
}

module.exports = {
  TerminalTranscriptBuffer
};

File: Sources/ClaudeWorkspace/Services/TerminalTranscriptBuffer.swift
import Foundation

final class TerminalTranscriptBuffer {
    private let maxLength: Int
    private var committedText: String
    private var currentLine: [Character]
    private var cursorColumn: Int

    init(seedText: String = "", maxLength: Int = 160_000) {
        self.maxLength = maxLength
        committedText = ""
        currentLine = []
        cursorColumn = 0
        replaceSeed(with: seedText)
    }

    var renderedText: String {
        committedText + String(currentLine)
    }

    func consume(_ rawText: String) -> String {
        apply(rawText)
        trimIfNeeded()
        return renderedText
    }

    func replaceSeed(with text: String) {
        if let newlineIndex = text.lastIndex(of: "\n") {
            let nextIndex = text.index(after: newlineIndex)
            committedText = String(text[..<nextIndex])
            currentLine = Array(text[nextIndex...])
        } else {
            committedText = ""
            currentLine = Array(text)
        }

        cursorColumn = currentLine.count
    }

    static func visibleText(in rawText: String) -> String {
        TerminalTranscriptBuffer(seedText: "", maxLength: max(rawText.count * 2, 4_096))
            .consume(rawText)
    }

    private func apply(_ rawText: String) {
        let scalars = Array(rawText.unicodeScalars)
        var index = 0

        while index < scalars.count {
            let scalar = scalars[index]

            if scalar == "\u{001B}" {
                index = parseEscapeSequence(in: scalars, startingAt: index)
                continue
            }

            switch scalar {
            case "\n":
                committedText.append(contentsOf: currentLine)
                committedText.append("\n")
                currentLine.removeAll(keepingCapacity: true)
                cursorColumn = 0
            case "\r":
                cursorColumn = 0
            case "\u{0008}", "\u{007F}":
                guard cursorColumn > 0 else {
                    break
                }

                cursorColumn -= 1

                if cursorColumn < currentLine.count {
                    currentLine.remove(at: cursorColumn)
                }
            case "\t":
                write(character: "\t")
            default:
                guard !CharacterSet.controlCharacters.contains(scalar) else {
                    break
                }

                write(character: Character(scalar))
            }

            index += 1
        }
    }

    private func parseEscapeSequence(in scalars: [UnicodeScalar], startingAt startIndex: Int) -> Int {
        let nextIndex = startIndex + 1

        guard nextIndex < scalars.count else {
            return startIndex + 1
        }

        switch scalars[nextIndex] {
        case "[":
            return parseCSISequence(in: scalars, startingAt: nextIndex + 1)
        case "]":
            return skipOSCSequence(in: scalars, startingAt: nextIndex + 1)
        default:
            return nextIndex + 1
        }
    }

    private func parseCSISequence(in scalars: [UnicodeScalar], startingAt startIndex: Int) -> Int {
        var index = startIndex
        var parameterScalars: [UnicodeScalar] = []

        while index < scalars.count {
            let scalar = scalars[index]

            if scalar.value >= 0x40, scalar.value <= 0x7E {
                applyCSI(final: scalar, parameterString: String(String.UnicodeScalarView(parameterScalars)))
                return index + 1
            }

            parameterScalars.append(scalar)
            index += 1
        }

        return index
    }

    private func skipOSCSequence(in scalars: [UnicodeScalar], startingAt startIndex: Int) -> Int {
        var index = startIndex

        while index < scalars.count {
            let scalar = scalars[index]

            if scalar == "\u{0007}" {
                return index + 1
            }

            if scalar == "\u{001B}", index + 1 < scalars.count, scalars[index + 1] == "\\" {
                return index + 2
            }

            index += 1
        }

        return index
    }

    private func applyCSI(final: UnicodeScalar, parameterString: String) {
        let parameters = parameterString
            .split(separator: ";", omittingEmptySubsequences: false)
            .map { Int($0) ?? 0 }

        switch final {
        case "m":
            return
        case "K":
            clearLine(mode: parameters.first ?? 0)
        case "J":
            clearScreen(mode: parameters.first ?? 0)
        case "C":
            moveCursorForward(by: max(parameters.first ?? 1, 1))
        case "D":
            moveCursorBackward(by: max(parameters.first ?? 1, 1))
        case "G":
            moveCursorToColumn(max((parameters.first ?? 1) - 1, 0))
        case "P":
            deleteCharacters(count: max(parameters.first ?? 1, 1))
        case "X":
            eraseCharacters(count: max(parameters.first ?? 1, 1))
        default:
            return
        }
    }

    private func write(character: Character) {
        if cursorColumn < currentLine.count {
            currentLine[cursorColumn] = character
        } else {
            while currentLine.count < cursorColumn {
                currentLine.append(" ")
            }

            currentLine.append(character)
        }

        cursorColumn += 1
    }

    private func clearLine(mode: Int) {
        switch mode {
        case 1:
            let upperBound = min(cursorColumn, currentLine.count)
            guard upperBound > 0 else {
                return
            }

            for index in 0..<upperBound {
                currentLine[index] = " "
            }
        case 2:
            currentLine.removeAll(keepingCapacity: true)
            cursorColumn = 0
        default:
            guard cursorColumn < currentLine.count else {
                return
            }

            currentLine.removeSubrange(cursorColumn...)
        }
    }

    private func clearScreen(mode: Int) {
        guard mode == 2 || mode == 3 else {
            return
        }

        committedText = ""
        currentLine.removeAll(keepingCapacity: true)
        cursorColumn = 0
    }

    private func moveCursorForward(by amount: Int) {
        cursorColumn += amount
    }

    private func moveCursorBackward(by amount: Int) {
        cursorColumn = max(cursorColumn - amount, 0)
    }

    private func moveCursorToColumn(_ column: Int) {
        cursorColumn = max(column, 0)
    }

    private func deleteCharacters(count: Int) {
        guard cursorColumn < currentLine.count else {
            return
        }

        let upperBound = min(cursorColumn + count, currentLine.count)
        currentLine.removeSubrange(cursorColumn..<upperBound)
    }

    private func eraseCharacters(count: Int) {
        guard cursorColumn < currentLine.count else {
            return
        }

        let upperBound = min(cursorColumn + count, currentLine.count)

        for index in cursorColumn..<upperBound {
            currentLine[index] = " "
        }
    }

    private func trimIfNeeded() {
        guard renderedText.count > maxLength else {
            return
        }

        replaceSeed(with: String(renderedText.suffix(maxLength)))
    }
}
